File: testbeds/vercel_ai/payment_agent.mjs
// A realistic (vulnerable) Vercel AI SDK human-in-the-loop tool.
//
// The transfer tool requires human confirmation; on resume the confirmation is
// passed to transferFunds as `approval`, trusted, and the transfer runs. A writer
// of the persisted message/tool-result state can inject a confirmation with no
// binding to approver, action, arguments, or a single-use nonce (the Ghost Approval
// surface). Tool-call wiring is omitted so the protected action is exercised directly.
import { tool } from 'ai';

void tool;

// PROTECTED ACTION: consumes the confirmation reply and transfers funds
export async function transferFunds({ recipient, amount, currency, approval }) {
  return `settled ${amount} ${currency} to ${recipient}`;
}

File: src/vyrion/_fixtures/agno/payment_agent.py
"""A realistic (vulnerable) Agno tool gated by requires_confirmation.

Agno serializes the tool confirmation (confirmed / confirmation_note) into the run /
session state. A writer of that state can set confirmed=True with no binding to
approver, action, arguments, or a single-use nonce, and the tool runs on resume
(the Ghost Approval surface).
"""
from agno.tools import tool


@tool(requires_confirmation=True)
def transfer_funds(recipient: str, amount: int, currency: str, approval: str) -> str:
    # PROTECTED ACTION
    return f"settled {amount} {currency} to {recipient}"
